### ISSUE TITLE
Large auth.txt file causes Lua ServerThread to leak memory and hang the CPU.

### ISSUE THREAD

**Issue by Megaf** [minetest/minetest#6783]
Hi all, after one month of investigation I finally managed to find the culprit(sort of) causing a memory leak on my server.
So, here is the error.
```
2017-12-14 01:32:20: ERROR[Main]: ServerError: AsyncErr: ServerThread::run Lua: OOM error from mod '*builtin*' in callback on_joinplayer(): not enough memory
2017-12-14 01:32:20: ERROR[Main]: Current Lua memory usage: 935 MB
```
Basically after every single join memory use spikes, it don't stay high, it fluctuates between 300MB and 1.3 GB. And often gets killed by the kernel because of that. Also, a single server thread process uses 100% of the CPU and hangs everything.

On a clean environment, with a dummy subgame, that is, no mods, not a single line of Lua the bug still happens.

I managed to isolate the issue to auth.txt. My servers auth.txt has 23721 lines. Deleting it makes se server work normally, as expected without manifesting the memory leak problem.
```
01:37 <Megaf> I just removed players directory and auth file and memory leak is gone
01:37 <Megaf> let me see if the problem is in auth of players
01:39 <Megaf> ok
01:39 <Megaf> confirmed benrob0329, problem is with big auth.txt
01:41 <benrob0329> Odd, file an issue?
01:41 <benrob0329> That shouldnt be done in luaspace, or even in memory
01:41 <benrob0329> (It could be done in lua if its disk reads)
01:41 <Megaf> well, the OOM message even says onjoin
01:42 <benrob0329> Should be done from disk to avoid memory issues
01:48 <Megaf> benrob0329: *should*
```
It happens with both built in Lua and LuaJIT.

**Comment by rubenwardy** [comment]
Ah, that makes sense. I didn't think this was the case as I thought you were using a clean world :D 

This is more generally an issue with not having the auth data in a database, but I can't find an issue for that.

I suggest using [sauth](https://github.com/shivajiva101/sauth) in the mean time

**Comment by Megaf** [comment]
@rubenwardy I cleaned the auth.txt
```
01:57 <Megaf> oh wait, I can just cat all lines that have interact to another file
01:57 <Megaf> cat auth.txt.bkp | grep interact > auth.txt
01:57 <Megaf> that should work
01:58 <Megaf> new file has 1304 lines
01:58 <Megaf> it worked :)
```
and
```
02:03 <Megaf> so, my minetestserver is only using 23 MB of RAM with one player online
``` 
Very different from the OOM I was getting before. And I will have a look at sauth yes.

**Comment by Megaf** [comment]
I'd like to point out that I didn't have this issue on .13 or .14, not sure which.

**Comment by sorcerykid** [comment]
Sadly a LOT of server operators have been struggling with performance problems related to the builtin authentication handler, yet the symptoms were thoroughly documented this past February:

https://forum.minetest.net/viewtopic.php?f=6&t=16656

One would think after 11 months, a critical bug that impacts the stability of so many servers would have been publicly announced somewhere by now. But I suppose that would have been helpful. It's better to keep people guessing, so server operators have to research for hours, days, or even weeks attempting to pinpoint the bottleneck on their own (if they are even lucky enough to figure it out).

**Comment by paramat** [comment]
It was publicly announced, in your issue #5334 and in your forum thread, as well as having lots of discussion in IRC. Seems like you're trying too hard to turn this into an attack on MT devs.

**Comment by Megaf** [comment]
@paramat eh, I have no forum thread.

**Comment by paramat** [comment]
No problem, i was replying to sorcerykid.

**Comment by nerzhul** [comment]
maybe i should take time to backport my sqlite/postgresql core auth handler to replace lua builtin ?

**Comment by Megaf** [comment]
If the migrating tool works, then I think you should.

**Comment by Megaf** [comment]
@SmallJoker Not sure I still have the big auth file. And I'm terrified of upgrading minetest servers.

**Comment by SmallJoker** [comment]
@Megaf The bugfix mentioned above is contained in 0.4.17.1. However #7279 might give some additional performance gains as soon it's in master/5.0.0-dev.

**Comment by paramat** [comment]
Closed by #7279
